contracts/modules/cover/CoverOrder.sol

```diff
@@ -0,0 +1,225 @@
+  /// @param limitOrderId The ID of the limit order
+  /// @param signature The signature of the order
+  function executeOrder(
+    BuyCoverParams calldata params,
+    PoolAllocationRequest[] calldata poolAllocationRequests,
+    uint limitOrderId,
+    bytes calldata signature
+  ) external payable returns (uint coverId) {
+
+    if (params.owner == address(0) || params.owner == address(this)) {
+      revert InvalidOwnerAddress();
+    }
+
+    if (params.paymentAsset == NXM_ASSET_ID) {
+      revert InvalidPaymentAsset();
+    }
+
+    // Compute hash
+    bytes32 orderHash = keccak256(
+      abi.encode(
+        params.productId,
+        params.coverId,
+        params.amount,
+        params.period,
+        params.paymentAsset,
+        params.coverAsset,
+        params.owner,
+        limitOrderId
+      )
+    );
+
+    // Ensure the order has not already been executed
+    if (executedOrders[orderHash]) {
+      revert OrderAlreadyExecuted(limitOrderId);
+    }
+
+    // Recover the signer address from the signature
+    bytes32 ethSignedOrderHash = orderHash.toEthSignedMessageHash();
+    address signer = ethSignedOrderHash.recover(signature);
+
+    // Ensure the signer is the user who owns the order
+    if (signer != params.owner) {
+      revert InvalidSignature();
+    }
+
+    // Mark the order as executed
+    executedOrders[orderHash] = true;
+
+    // ETH payment
+    if (params.paymentAsset == ETH_ASSET_ID) {
+      return _buyCoverEthPayment(params, poolAllocationRequests);
+    }
+
+    // ERC20 payment
+    return _buyCoverErc20Payment(params, poolAllocationRequests);
+  }
+
+  /// @notice Handles ETH/WETH payments for buying cover.
+  /// @dev Transfers WETH tokens from the order creator to the contract, then unwraps it,  then buys cover on behalf of the creator.
+  //       Calculates ETH refunds if any and sends back to param.owner.
+  /// @param params The parameters required to buy cover.
+  /// @param poolAllocationRequests The allocation requests for the pool's liquidity.
+  /// @return coverId The ID of the purchased cover.
+  function _buyCoverEthPayment(
+    BuyCoverParams calldata params,
+    PoolAllocationRequest[] calldata poolAllocationRequests
+  ) internal returns (uint coverId) {
+
+    uint ethBalanceBefore = address(this).balance;
+
+    weth.transferFrom(params.owner, address(this), params.maxPremiumInAsset);
+    weth.withdraw(params.maxPremiumInAsset);
+
+    coverId = cover.buyCover{value: params.maxPremiumInAsset}(params, poolAllocationRequests);
+
+    uint ethBalanceAfter = address(this).balance;
+
+    // transfer any ETH refund back to params.owner
+    if (ethBalanceAfter > ethBalanceBefore) {
+      uint ethRefund = ethBalanceAfter - ethBalanceBefore;
+      (bool sent,) = payable(params.owner).call{value: ethRefund}("");
+      if (!sent) {
+        revert TransferFailed(params.owner, ethRefund, ETH);
+      }
+    }
+
+    // Emit event
+    emit OrderExecuted(params.owner, coverId);
```

Comment:
Sync with Tomasz regarding what info we'd want to include in the events because I bet he'd want more info here.